### PR TITLE
Port to python 3

### PR DESCRIPTION
**Pull request by mjms3**
Deliberately unsquashed at this point as I thought it might be helpful to leave the individual commits.

Not very familiar with travis build stages so not sure if the config is the best way to achieve this.

Large change to doctests format is due to the ordering of the email.as_string() method not being consistent going from p2 -> p3. In doing this, I noticed that the previous examples were missing the X-Log-Level header due to the doctests ellipsis matching too much. The previous method of testing with printing to screen meant that the tests of emails that aren't sent did something meaningful. With my new implementation, not so much, but I wasn't sure it was particularly worth the effort of adding an explicit check for this in the doctests as I don't think it adds to the clarity of the documentation.

Also tested manually locally with utf-8 encoded log messages and it all appears to work correctly.

Obviously happy to discuss / revisit / revise.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -23,7 +33,7 @@ jobs:
       after_success: skip
 
       install: "pip install -U coveralls-check"
-      script: "coveralls-check $TRAVIS_COMMIT"
+      script: "coveralls-check --fail-under=95 $TRAVIS_COMMIT"
```
[the remark this answers is not in the document]
Comment (mjms3) [reply .travis.yml]: https://coveralls.io/builds/16407410. When I ran this build, the coveralls check only picked up the coverage from the 2.7 build and didn't combine it with the 3.6 build, so I assumed that it couldn't combine the coverage across the different two different builds. It appears that it did after this. Really not sure what's going on here but obviously more than happy to remove it and go back to 100%

[the remark this answers is not in the document]
Comment (mjms3) [reply .travis.yml]: (https://travis-ci.org/mjms3/mailinglogger/builds/363837040)

File: .travis.yml [hunk 2]
```diff
@@ -14,6 +14,16 @@ jobs:
...
+      python: 2.7
+      after_success:
+        - coveralls
+
+    - stage: test
+      install:
+      - "pip install --upgrade pip setuptools"
+      - "pip install -Ue .[test]"
+      script: coverage run --source mailinglogger -m pytest
+      python: 3.6
```
[the remark this answers is not in the document]
Comment (mjms3) [reply .travis.yml]: I tried that: https://travis-ci.org/mjms3/mailinglogger/builds/363778602 and it looked really weird. I probably did something silly though

File: docs/summarisinglogger.txt [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
-  :mod:`smtplib` are printed to the screen so we can see what's going on:
+  :mod:`smtplib` are checked against the expected message so that we can
+  see what's going on:
 
   >>> import smtplib
   >>> server = smtplib.SMTP('localhost')
   >>> server.sendmail('from@example.com', ['to@example.com'], 'The message')
-  sending to ['to@example.com'] from 'from@example.com' using ('localhost', 25)
-  The message
+  >>> server.check_sent_message_matches("""\
```
[the remark this answers is not in the document]
Comment (mjms3) [reply docs/summarisinglogger.txt]: I'll discuss this in my response to your question aboutcheck_sent_message_matches

File: docs/summarisinglogger.txt [hunk 4]
```diff
@@ -64,35 +66,36 @@ when the script that calls the logging function exits.
...
 
 The logging on script exit is done using python's :mod:`atexit`
 module. We can see that the handler we created above is now
 registered with this module by having a poke at its internals:
 
->>> import atexit
->>> print atexit._exithandlers
-[(<bound method SummarisingLogger.close of <...>, (), {})]
+>>> from mailinglogger.common import exit_handler_manager
+>>> print(exit_handler_manager._AT_EXIT_HANDLERS)
```
[the remark this answers is not in the document]
Comment (mjms3) [reply docs/summarisinglogger.txt]: The API is only present in python 2.x. In python 3.x, atexit handlers are all implemented in C so I don't think the API exists but I could be wrong. Would love to hear your thoughts

File: mailinglogger/common.py [hunk 5]
```diff
@@ -2,6 +2,31 @@
...
+
+    def clear_at_exit_handlers(self):
+
+        if hasattr(atexit,'unregister'):
+            [atexit.unregister(f) for f in self._AT_EXIT_HANDLERS]
+        else:
+            atexit._exithandlers[:] = []
+        self._AT_EXIT_HANDLERS = []
+
+exit_handler_manager = AtExitHandlerManager()
```
[the remark this answers is not in the document]
Comment (mjms3) [reply mailinglogger/common.py]: I'm really really not a fan of the whole approach (having the manager, using the global state etc). the exit_handler_manager = AtExitHandlerManager() line definitely isn't needed. Again would love to hear your thoughts. My main aim was to try to maintain the spirit of the documentation that allowed you to see what was registered.

File: setup.py [hunk 6]
```diff
@@ -28,6 +28,6 @@
     include_package_data=True,
     extras_require=dict(
         test=['coverage', 'coveralls', 'pytest', 'sybil', 'testfixtures'],
-        build=['sphinx', 'pkginfo', 'setuptools-git', 'wheel', 'twine'],
+        build=['sphinx', 'pkginfo', 'setuptools-git', 'wheel', 'twine', 'six'],
```
[the remark this answers is not in the document]
Comment (mjms3) [reply setup.py]: Yes. This is my bad. I got confused about how everything was configured.

We could have a compat.p; however, I don't think that a dependency on six,py is particularly onerous. Happy to change it if you'd prefer though.

**Comment by mjms3** [comment]
* what's driving check_sent_message_matches? I'd ideally prefer to keep the assertions as they were, so I'm trying to understand the reason for the change? edit: Is this the email.as_string() ordering thing?
** Yes, this is because of the the email.as_string() ordering thing + the fact that the tests as they were were quite brittle (and actually were passing incorrectly due to over enthusiastic matching by the doctest ellipsis). However, I'm really, really, really not keen on the way they made the documentation look. I was thinking about this and think that the best way to achieve the same documentation look as before but to have more ability to compare the actual to the expected ie to implement the check_sent_messages style check (or possibly something a bit neater and less bleurgh) would be to wrap the expected results in a block with a custom Parser and still print to screen-- i'm not that familiar with all the tooling that you've used, but I think that shouldn't be too hard to achieve. Thoughts?

Would it be simpler to just make a 5.x that was Python 3 only? The code changes so infrequently that I reckon maintaining a 4.x for Python 2 and having 5.x be Python 3 only.
** It's your project, so I'll leave you to make that call but I'd personally prefer to have code that is compatible with python 2 and 3... the reason I originally did the port is that I wanted to submit a PR for some different functionality and I'd want that on both python 2 and python 3.

Thanks for the quick response and questions

**Comment by mjms3** [comment]
Follow up. 

As a side note, Sybil is awesome!!

Anyhow, I think that the cleanest thing to do with respect to the tests is to create a custom OutputChecker that can be passed to the DocTestRunner. Thoughts?

This will either mean creating a new DocTestParser specific to the MailingLogger, or modifying the one in sybil (along with the Sybil DocTestRunner) to allow the `checker` option to be passed through.

Sound reasonable?

File: .travis.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (cjw296) [reply .travis.yml]: Coveralls aggregation (on occasion) takes longer to accumulate coverage results than the build is prepared to wait, or they report <100% even though all the results aren't in. It's frustrating but I haven't found a better alternative...

File: .travis.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjw296) [reply .travis.yml]: Here's the kind of thing we should end up with: https://github.com/cjw296/sybil/blob/master/.travis.yml

File: mailinglogger/common.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cjw296) [reply mailinglogger/common.py]: Here's how I tackled this in another project, maybe you could do something similar here:
https://github.com/Simplistix/testfixtures/commit/68f9e17bd51bf58eb860a1c832cca9d784e55d51#diff-9afec5079d8942d01e2e06ac84630c44R306

File: setup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cjw296) [reply setup.py]: Let's go with six but please put it in install_requires where it belongs ;-)

**Comment by cjw296** [comment]
Glad you like Sybil, @benji-york is the genius behind it with Manuel, I just slimmed Manuel down so it would fit in my brain.

So, I think a new sybil parser is the way to go here. We need to find a ReStructuredText rendering of the mail messages that both looks nice and is easily parseable.

I'm not a fan of wiring in any more through the DocTestParser, doctests are convoluted and unmaintained in their implementation and Sybil gives us a nicer way to go.

Good luck, happy to help where I can but I'm seriously limited on time.

Also curious: what was the different functionality you're looking to implement? (maybe stick it in another issue for discussion...)

**Comment by cjw296** [comment]
Hey @mjms3 :-) Any joy with this? I'm likely to need it myself soon so wanted to see how things were going...

**Comment by mjms3** [comment]
Hey @cjw296. Afraid I haven't had a chance to look at it much since *embarrassed face*. I was planning to spend some time on it over the long weekend, so hopefully I'll make some meaningful progress then :D.

**Comment by cjw296** [comment]
@mjms3 - I think #14 is ready to land now, would you mind taking a look?
I'll close this PR now the other one is ready...

**Pull request closed without merging**